rust/private/rust.bzl

```diff
@@ -133,7 +133,7 @@ def _rust_binary_impl(ctx):
     if (toolchain.target_arch == "wasm32"):
         output = ctx.actions.declare_file(ctx.label.name + ".wasm")
     else:
-        output = ctx.actions.declare_file(ctx.label.name)
+        output = ctx.actions.declare_file(ctx.label.name.replace("-", "_"))
```

Comment:
Is this correct? I believe binary names somewhat commonly contain hyphens

Reply:
Also suspect that wasm scheme is different

Reply:
> I believe binary names somewhat commonly contain hyphens

That may be the case, but previously rust_binary with a hyphen didn't work at all:

```bazel
rust_binary(
    name = "test-hyphen",
    srcs = ["main.rs"],
)
```

```console
$ bazel run test-hyphen --verbose_failures
INFO: Writing tracer profile to '/home/david/.cache/bazel/_bazel_david/ebce1d0721fb68dda9c70c0dd1405803/command.profile.gz'
INFO: Analyzed target //:test-hyphen (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
ERROR: /git/cxx/BUILD:3:1: error executing shell command: '/bin/bash -c CARGO_MANIFEST_DIR=$(pwd)/BUIL external/rust_1_42_beta/bin/rustc "$@" --remap-path-prefix="$(pwd)"=__bazel_redacted_pwd  main.rs --crate-name=test-hyphen --crate-type=bin --codegen=met...' failed (Exit 1) bash failed: error executing command 
  (cd /home/david/.cache/bazel/_bazel_david/ebce1d0721fb68dda9c70c0dd1405803/sandbox/linux-sandbox/1167/execroot/__main__ && \
  exec env - \
    CARGO_CFG_TARGET_ARCH=x86_64 \
    CARGO_CFG_TARGET_OS=linux \
    CARGO_PKG_AUTHORS='' \
    CARGO_PKG_DESCRIPTION='' \
    CARGO_PKG_HOMEPAGE='' \
    CARGO_PKG_NAME=test-hyphen \
    CARGO_PKG_VERSION=0.0.0 \
    CARGO_PKG_VERSION_MAJOR=0 \
    CARGO_PKG_VERSION_MINOR=0 \
    CARGO_PKG_VERSION_PATCH=0 \
    CARGO_PKG_VERSION_PRE='' \
  /bin/bash -c 'CARGO_MANIFEST_DIR=$(pwd)/BUIL external/rust_1_42_beta/bin/rustc "$@" --remap-path-prefix="$(pwd)"=__bazel_redacted_pwd' '' main.rs '--crate-name=test-hyphen' '--crate-type=bin' '--codegen=metadata=' '--out-dir=bazel-out/k8-fastbuild/bin' '--codegen=extra-filename=' '--codegen=opt-level=0' '--codegen=debuginfo=0' '--emit=dep-info,link' '--color=always' '--target=x86_64-unknown-linux-gnu' -L external/rust_1_42_beta/lib/rustlib/x86_64-unknown-linux-gnu/lib '--edition=2018' '--codegen=linker=/usr/bin/gcc' --codegen 'link-args=-fuse-ld=gold -Wl,-no-as-needed -Wl,-z,relro,-z,now -B/usr/bin -pass-exit-codes -lstdc++ -lm')
Execution platform: @local_config_platform//:host

Use --sandbox_debug to see verbose messages from the sandbox
error: invalid character `-` in crate name: `test-hyphen`

error: aborting due to previous error

Target //:test-hyphen failed to build
INFO: Elapsed time: 0.140s, Critical Path: 0.09s
INFO: 0 processes.
FAILED: Build did NOT complete successfully
FAILED: Build did NOT complete successfully
```

After this PR rustc will see it as if the user had written the name with an underscore, but the BUILD file and bazel command line will be allowed to use hyphen to match the actual name upstream on crates.io.

I believe it would be a separate feature request to pass the right `-o` flag to rustc to write out the binary with an underscore in the filename.

Reply:
It's strange behavior in rustc I guess, fair enough.